File: src/main/angular/model/recommendation_action.ts
/** Represents a single action that a recommendation implemented. */
export class RecommendationAction {
  constructor(
    private affectedAccount: string,
    private previousRole: string,
    private newRole: string
  ) {}

  /** Converts the given action to a string that is visible to the user. */
  public toString(): string {
    if (this.newRole.length > 0) {
      // Role was replaced
      return `Replace ${this.previousRole} with ${this.newRole} on ${this.affectedAccount}.`;
    }
    // Role was removed
    return `Remove ${this.previousRole} from ${this.affectedAccount}.`;
  }
}

File: src/main/java/com/google/impactdashboard/data/recommendation/RecommendationAction.java
package com.google.impactdashboard.data.recommendation;

import com.google.auto.value.AutoValue;

/** Represents an action taken by a GCP Recommendation. */
@AutoValue
public abstract class RecommendationAction {

  public abstract String getAffectedAccount();
  public abstract String getPreviousRole();
  public abstract String getNewRole();
  public abstract ActionType getActionType();

  public enum ActionType {
    REMOVE_ROLE,
    REPLACE_ROLE
  }

  /** 
   * Creates a {@code RecommendationAction} representing an action of type 
   * {@code actionType} taken by a GCP Recommendation that changed the role of 
   * user {@code affectedAccount} from {@code previousRole} to {@code newRole}. 
   * If the user's role was removed instead of replaced, then {@code newRole} 
   * should be the empty string. 
   */
  public static RecommendationAction create(String affectedAccount, String previousRole, 
    String newRole, ActionType actionType)  {
    return new AutoValue_RecommendationAction(affectedAccount, previousRole, newRole, actionType);
  }
}
